core/cache/resident.go

```diff
@@ -79,8 +86,43 @@ func (m *residentManager) new() *Resident {
 	return r
 }
 
-func (m *residentManager) evict(id uint64) {
-	// TODO (manadart 2019-04-17): TBC when the mark/sweep logic is added.
+// mark sets all of the manager's residents to be stale.
+func (m *residentManager) mark() {
+	for _, r := range m.residents {
+		r.stale = true
+	}
+}
+
+// sweep removes stale cache residents in descending order of ID.
+// Because the IDs are supplied in increasing order, this ensures
+// we never remove a resident's model before the resident itself.
+func (m *residentManager) sweep() error {
+	// Create a descending order slice of IDs.
+	residentIds := make([]uint64, len(m.residents))
+	i := 0
+	for id := range m.residents {
+		residentIds[i] = id
+		i++
+	}
+	sort.Sort(uint64Reverse(residentIds))
+
+	// Read the map "leaves-first" and evict stale residents.
```

Comment:
LIFO (last in first out) is probably more recognizable than "leaves-first"

Reply:
The model doesn't ensure an application exists before it updates the unit. Is this going to cause a problem?

Reply:
No. The only concern is the way removal is delegated to the model for all other resident types. 

If we remove a model in the sweep before its children, then those won't have their resources released and be de-registered.